### PR TITLE
apiextensions: fix items+type logic in API due to broken go-openapi validation

### LINKED ISSUE #84880: Property names "items" in OpenAPISchema for CRD not handled correctly
**What happened**:
 Created a CRD with OpenAPISchema containing an array which itself contains objects which have a property names "items". This already exists in [DownwardAPI volume source](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#downwardapivolumesource-v1-core).

Creating a resource with this CRD fails with:
`The Foo "test" is invalid: type: Required value`

**What you expected to happen**:

It would succeed CRD validation.

**How to reproduce it (as minimally and precisely as possible)**:
kind create cluster --image kindest/node:v1.16.2
Kubectl apply the following CRD:

```
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: foos.stable.example.com
spec:
  group: stable.example.com
  versions:
  - name: v1
    served: true
    storage: true
  scope: Namespaced
  names:
    plural: foos
    singular: foo
    kind: Foo
  preserveUnknownFields: false
  validation:
    openAPIV3Schema:
      type: object
      properties:
        apiVersion:
          type: string
        kind:
          type: string
        metadata:
          type: object
        spec:
          type: object
          properties:
            volumes:
              type: array
              items:
                type: object
                properties:
                  downwardAPI:
                    properties:
                      items:
                        items:
                          properties:
                            path:
                              type: string
                          required:
                          - path
                          type: object
                        type: array
                    type: object
```

Then kubectl apply the following

```
apiVersion: stable.example.com/v1
kind: Foo
metadata:
  name: test
spec:
  volumes:
  - downwardAPI:
      items:
        - path: annotations
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.2", GitCommit:"c97fe5036ef3df2967d086711e6c0c405941e14b", GitTreeState:"clean", BuildDate:"2019-10-15T19:18:23Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.2", GitCommit:"c97fe5036ef3df2967d086711e6c0c405941e14b", GitTreeState:"clean", BuildDate:"2019-11-05T23:58:50Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
```

- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
```
NAME="Linux Mint"
VERSION="18.3 (Sylvia)"
ID=linuxmint
ID_LIKE=ubuntu
PRETTY_NAME="Linux Mint 18.3"
VERSION_ID="18.3"
HOME_URL="http://www.linuxmint.com/"
SUPPORT_URL="http://forums.linuxmint.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/linuxmint/"
VERSION_CODENAME=sylvia
UBUNTU_CODENAME=xenial
```
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

This issue was meant to be fixed by https://github.com/go-openapi/validate/pull/107
However, there is a discussion: https://github.com/go-openapi/validate/issues/108#issuecomment-540532114

### PR DESCRIPTION
**Pull request by sttts**
* Half of the fix (items without type) was done in https://github.com/kubernetes/kubernetes/pull/76124.
* Second half (type:array without items) is done here.
* In addition this fixes 
* Finally we forgot to disable go-openapi meta checks for default validation.

Plus this PR add an integration test.

/kind bug

Fixes  https://github.com/kubernetes/kubernetes/issues/84880

```release-note
CRDs can have fields named `type` with value `array` and nested array with `items` fields without validation to fall over this.
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -472,6 +476,7 @@ replace (
 	github.com/xordataexchange/crypt => github.com/xordataexchange/crypt v0.0.3-0.20170626215501-b2862e3d0a77
 	go.etcd.io/bbolt => go.etcd.io/bbolt v1.3.3
 	go.etcd.io/etcd => go.etcd.io/etcd v0.0.0-20191023171146-3cf2f69b5738
+	go.mongodb.org/mongo-driver => go.mongodb.org/mongo-driver v1.1.2
```
Comment (liggitt) [inline go.mod]: what pulled this in?

File: staging/src/k8s.io/api/go.mod [hunk 2]
```diff
@@ -6,11 +6,14 @@ go 1.12
...
 	github.com/gogo/protobuf v1.2.2-0.20190723190241-65acae22fc9d
-	github.com/stretchr/testify v1.3.0
+	github.com/stretchr/testify v1.4.0
 	k8s.io/apimachinery v0.0.0
 )
 
 replace (
+	github.com/go-openapi/spec => github.com/go-openapi/spec v0.19.2
+	github.com/mailru/easyjson => github.com/mailru/easyjson v0.0.0-20190626092158-b2ccc519800e
+	github.com/stretchr/testify => github.com/stretchr/testify v1.3.0
```
Comment (liggitt) [inline staging/src/k8s.io/api/go.mod]: need to satisfy `hack/lint-dependencies.sh` so our pinned versions match naturally preferred versions

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 3]
```diff
@@ -713,7 +713,7 @@ func (r *crdHandler) getOrCreateServingInfoFor(uid types.UID, name string) (*crd
 					if err := apiservervalidation.ConvertJSONSchemaProps(&statusSchema, openapiSchema); err != nil {
 						return nil, err
 					}
-					statusValidator = validate.NewSchemaValidator(openapiSchema, nil, "", strfmt.Default, validate.DisableObjectArrayTypeCheck(true))
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: I'm confused... I thought we wanted this option ("we forgot to disable go-openapi meta checks for default validation.")

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go [hunk 4]
```diff
@@ -39,7 +39,7 @@ func NewSchemaValidator(customResourceValidation *apiextensions.CustomResourceVa
 			return nil, nil, err
 		}
 	}
-	return validate.NewSchemaValidator(openapiSchema, nil, "", strfmt.Default, validate.DisableObjectArrayTypeCheck(true)), openapiSchema, nil
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go]: I'm confused... I thought we wanted this option ("we forgot to disable go-openapi meta checks for default validation.")

**Comment by josiahbjorgaard** [comment]
/milestone v1.18

(code freeze v1.17)

File: go.mod (hunk 1 above)
Comment (sttts) [reply go.mod]: Only called pin-dependencies and update-vendor. But interesting indeed.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go (hunk 3 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: The option was switch from Disable to Enable. The default is now sane.

**Comment by liggitt** [comment]
/milestone v1.17
release blocking bug

File: go.mod (hunk 1 above)
Comment (sttts) [reply go.mod]: `go-openapi/strfmt`.  It got support for Mongo's bson format. 

We should immediately start restricting the supported formats to what we have in 1.16. I fear we get more and more format without noticing.

File: staging/src/k8s.io/api/go.mod (hunk 2 above)
Comment (sttts) [reply staging/src/k8s.io/api/go.mod]: done

File: go.mod (hunk 1 above)
Comment (liggitt) [reply go.mod]: Agree. This is concerning, both from what it exposes in our API, and what it pulls in as dependencies. I think we need to open upstream issues requesting things like this be split so all downstreams don't pick up this dependency, or look into replacing this library long term.

Comment (liggitt) [reply go.mod]: > We should immediately start restricting the supported formats to what we have in 1.16.

do we restrict format today? or is it allowed to be arbitrary and some special values just get magical validation behavior from go-openapi?

even if we lock down our API surface, I don't want to pick up dependencies like this, so I don't think we should take this unless we have a plan for dropping it

Comment (smarterclayton) [reply go.mod]: Agreed.  We need a very tight control of public APIs, and do not want to be dependent on dependencies for api stability.

Comment (sttts) [reply go.mod]: https://github.com/kubernetes/kubernetes/pull/85223 as a first step.

Comment (sttts) [reply go.mod]: > do we restrict format today? or is it allowed to be arbitrary and some special values just get magical validation behavior from go-openapi?

By definition formats are best effort to be checked by OpenAPI implementors. We allow any format string. go-openapi checks two handful of format strings, see PR.

Comment (liggitt) [reply go.mod]: #85223 looks good for controlling our API surface.

Comment (liggitt) [reply go.mod]: @sttts pointed out that our openapi validation already required a mongo driver to do bsonobjectid validation (github.com/globalsign/mgo). this drops that unofficial mongo code (7771 lines worth) in favor of the official driver (13402 lines worth).

I'm ok with the following plan:
1. control api surface in #85223
2. merge this to fix the items validation bug
3. open issue upstream in openapi-validator/strfmt to see if they'll refactor to drop the mongo dep
4. open an issue in k/k consider ways to clean up and control our openapi validator dependency, even if it means extracting the narrow pieces we depend on

Comment (sttts) [reply go.mod]: github.com/globalsign/mgo replaced by go.mongodb.org/mongo-driver, the former unmaintained according to the go-openapi project.

**Comment by sttts** [comment]
@liggitt #85381 merged. Is this good to go?

**Comment by liggitt** [comment]
Dependency impact:
* drops github.com/globalsign/mgo in favor of go.mongodb.org/mongo-driver
* adds indirect dependencies on the following, but doesn't actually vendor any code from these:
  * github.com/tidwall/pretty (used by go-openapi/strfmt)
  * github.com/vektah/gqlparser (used by go-openapi/validate)
    * github.com/agnivade/levenshtein
    * github.com/andreyvit/diff
    * github.com/sergi/go-diff

the second set of dependencies doesn't impact our vendor dir, but does impact our dependency complexity and add to my opinion that we should look to trim down the go-openapi dependencies we have

File: vendor/github.com/go-openapi/analysis/mixin.go [hunk 5]
```diff
@@ -27,7 +27,7 @@ import (
 // collisions are avoided by appending "Mixin<N>" but only if
 // needed.
 //
-// The following parts of primary are never modified by merging:
+// The following parts of primary are subject to merge, filling empty details
```
Comment (liggitt) [inline vendor/github.com/go-openapi/analysis/mixin.go]: that's a terrifying godoc change for a point release. do we depend on the previous behavior?

Comment (liggitt) [reply vendor/github.com/go-openapi/analysis/mixin.go]: answering my own question, no calls to this in our codebase

File: vendor/github.com/go-openapi/strfmt/time.go [hunk 6]
```diff
@@ -56,12 +55,14 @@ const (
...
 	RFC3339Micro = "2006-01-02T15:04:05.000000Z07:00"
+	// ISO8601LocalTime represents a ISO8601 format to ISO8601 in local time (no timezone)
+	ISO8601LocalTime = "2006-01-02T15:04:05"
 	// DateTimePattern pattern to match for the date-time format from http://tools.ietf.org/html/rfc3339#section-5.6
 	DateTimePattern = `^([0-9]{2}):([0-9]{2}):([0-9]{2})(.[0-9]+)?(z|([+-][0-9]{2}:[0-9]{2}))$`
 )
 
 var (
-	dateTimeFormats = []string{RFC3339Micro, RFC3339Millis, time.RFC3339, time.RFC3339Nano}
+	dateTimeFormats = []string{RFC3339Micro, RFC3339Millis, time.RFC3339, time.RFC3339Nano, ISO8601LocalTime}
```
Comment (liggitt) [inline vendor/github.com/go-openapi/strfmt/time.go]: this adds to the accepted formats the `datetime` validator permits, the docs added in https://github.com/kubernetes/kubernetes/pull/85381/files#r347187064 are not complete

File: vendor/github.com/go-openapi/validate/schema_option.go [hunk 7]
```diff
@@ -14,20 +14,41 @@
 
 package validate
 
+// SchemaValidatorOptions defines optional rules for schema validation
 type SchemaValidatorOptions struct {
-	DisableObjectArrayTypeCheck bool
+	EnableObjectArrayTypeCheck    bool
+	EnableArrayMustHaveItemsCheck bool
```
Comment (liggitt) [inline vendor/github.com/go-openapi/validate/schema_option.go]: do we need to turn these on when validating the schema itself? does this change to default these off loosen our validation on CRD's themselves (not the CRs)?

**Comment by liggitt** [comment]
one question about CRD validation, and one comment about datetime validation (not sure what to do with that one), then lgtm

File: vendor/github.com/go-openapi/validate/schema_option.go (hunk 7 above)
Comment (sttts) [reply vendor/github.com/go-openapi/validate/schema_option.go]: We don’t use meta validation of schemas using go-openapi.

File: vendor/github.com/go-openapi/strfmt/time.go (hunk 6 above)
Comment (sttts) [reply vendor/github.com/go-openapi/strfmt/time.go]: we can easily override the time format, copy their regexps and that way stop adding ISO8601LocalTime.

Comment (sttts) [reply vendor/github.com/go-openapi/strfmt/time.go]: The validation code path does not use the `ISO8601LocalTime` regex, but only the one for rfc3339. This means that the `datetime` format in our API rejects e.g. timezone-less `2006-01-02T15:04:05`, but also the mille and micro second variant. I don't know why the upstream project added that support to their unmarshaller.

Comment (sttts) [reply vendor/github.com/go-openapi/strfmt/time.go]: With that, I think this PR is fine.

**Comment by liggitt** [comment]
can you squash these together:
apiextensions: disable meta checks in go-openapi for defaults
apiextensions: fixup go-openapi constructor invocations

and these together:
hack/pin-dependency.sh github.com/go-openapi/validate v0.19.5
Fixup pinned versions

File: vendor/github.com/go-openapi/strfmt/time.go (hunk 6 above)
Comment (liggitt) [reply vendor/github.com/go-openapi/strfmt/time.go]: can you open a followup issue to track the override/cleanup work we discussed:
* pinning the named/supported validators we currently have
* forking/extracting the openapi validator

**Comment by liggitt** [comment]
#85320 flake
/retest

**Pull request merged**